### PR TITLE
Feature/tao 3077 tooltip in textreader

### PR DESCRIPTION
**Pull request by no-chris**
depends on https://github.com/oat-sa/extension-tao-itemqti/pull/799

There is a test item in the folder : 
pciSamples/views/js/pciCreator/dev/textReaderInteraction/test/data
that can be useful for testing authoring

### CODE REVIEW

File: views/js/pciCreator/dev/textReaderInteraction/creator/js/userTooltips.js [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    $,
+    eventifier,
+    textWrapper,
+    markupTpl,
+    createButtonTpl,
+    authoringTpl
+) {
+    'use strict';
+
+    function buildId(tooltipsData) {
```
Comment (krampstudio) [inline views/js/pciCreator/dev/textReaderInteraction/creator/js/userTooltips.js]: What's the goal of this function ? If it's to create a unique identifier, better use `lib/uuid`

File: views/js/pciCreator/dev/textReaderInteraction/creator/tpl/userTooltips/authoring.tpl [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<hr />
+<h3>Tooltips</h3>
```
Comment (krampstudio) [inline views/js/pciCreator/dev/textReaderInteraction/creator/tpl/userTooltips/authoring.tpl]: Could you please use the translation tag.

File: views/js/pciCreator/dev/textReaderInteraction/test/test.html [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+<head>
+    <meta charset="utf-8">
+    <title>Text Reader Interaction Test</title>
+    <base href="../../../../../../../tao/views/" />
+
+    <link rel="stylesheet" type="text/css" href="js/lib/qunit/qunit.css">
+    <link rel="stylesheet" type="text/css" href="css/tao-main-style.css">
+    <script type="text/javascript" src="js/lib/require.js"></script>
+    <script type="text/javascript" src="js/lib/qunit/qunit.js"></script>
+    <!--<script type="text/javascript" src="js/lib/blanket/blanket.min.js" data-cover-only="/pciSamples/views/js/pciCreator/dev/textReaderInteraction/"></script>-->
```
Comment (krampstudio) [inline views/js/pciCreator/dev/textReaderInteraction/test/test.html]: any reason why the coverage isn't activated ?

File: views/js/pciCreator/dev/textReaderInteraction/test/test.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
+define([
```
Comment (krampstudio) [inline views/js/pciCreator/dev/textReaderInteraction/test/test.js]: Even on tests we need the copyright/license header

File: views/js/pciCreator/dev/textReaderInteraction/test/test.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    var runner;
+
+    //manually register the pci from its manifest
+    pciTestProvider.addManifestPath(
+        'textReaderInteraction',
+        'pciSamples/pciCreator/dev/textReaderInteraction/pciCreator.json');
+    ciRegistry.resetProviders();
+    ciRegistry.registerProvider(pciTestProvider.getModuleName());
+
+    module('Text Reader Interaction');
```
Comment (krampstudio) [inline views/js/pciCreator/dev/textReaderInteraction/test/test.js]: Should be `QUnit.module`

File: views/js/pciCreator/dev/textReaderInteraction/creator/js/userTooltips.js (hunk 1 above)
Comment (no-chris) [inline views/js/pciCreator/dev/textReaderInteraction/creator/js/userTooltips.js]: indeed, but to be on the safe side, I wouldn't change the rest of the code...

**Comment by no-chris** [comment]
All requested changes are done!

**Comment by krampstudio** [comment]
I was able to create tooltips successfully.

But following error appears in the console each time I click on a wrapped text in the authoring (I thought I need to click on the yellow area to add the content). The error doesn't seem to have a direct impact but this could hide some side effects: 

```
InvalidStateError: An attempt was made to use an object that is not, or is no longer, usable 
wrapSelectionhttps://taodev.localdomain/taoQtiItem/views/js/qtiCreator/widgets/helpers/textWrapper.js:41:17
textWrapper.create/<https://taodev.localdomain/taoQtiItem/views/js/qtiCreator/widgets/helpers/textWrapper.js:80:20
p.event.dispatchhttps://taodev.localdomain/tao/views/js/lib/jquery-1.8.0.min.js:2:37892
p.event.add/hhttps://taodev.localdomain/tao/views/js/lib/jquery-1.8.0.min.js:2:33864
```

There's aslo a 2 edge cases (check if you have to fix them or not) : 

 - Without content the tooltip appears but empty (check what we should do when empty).
 - I was able to insert HTML content inside tooltips. The content should be escaped in order to prevent XSS injections.

**Comment by krampstudio** [comment]
Also I'm wondering if the extension shouldn't have a direct dependency to taoQtiItem (to ensure the portable shared library is there). 
@ssipasseuth shouldn't we increase the version number for the PCI ? (I don't know how PCI versionning works)

**Comment by no-chris** [comment]
>Without content the tooltip appears but empty (check what we should do when empty).
The renderer doesn't initialize anymore empty tooltips. So we can still keep them as placeholders for later authoring, but they won't be rendered until some content is added to them.

other fixes are in.

**Comment by ssipasseuth** [comment]
Sorry, I really need to document this asap for everybody to know.

The version increase for the PCI should be done on the pciCreator.json. https://github.com/oat-sa/extension-pcisample/blob/develop/views/js/pciCreator/dev/textReaderInteraction/pciCreator.json
This looks like a backward-compatible feature so it should be intermediate version increase : 0.2.1 -> 0.3.0
As long as you don't need to migrate your item with an external php script, it will be either and intermediate or minor version bump.

And indeed there is some dependency to a specific taoQtiItem because of the shared lib, it is because we don't have a higher level qtiItemPortableElement extension to host all shared libraries... nice catch

**Comment by no-chris** [comment]
@ssipasseuth this page is dying for your documentation:
http://forge.taotesting.com/projects/tao/wiki/PCI_development
😄 

ok, I updated the PCI version and the taoQtiItem dependency requirement

**Pull request merged**